docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1058 @@
+
+### Scalar remainder handling
+
+Imagine that we want to calculate the sum of all the numbers in given buffer. We definitely want to add every element just once, without repetitions. That is why in the first loop, we add four (128 bits / 32 bits) integers in one iteration. In the second loop, we handle the remaining values.
+
+
+```cs
+int Sum(Span<int> buffer)
+{
+    Debug.Assert(Vector128.IsHardwareAccelerated && buffer.Length >= Vector128<int>.Count);
+
+    // The initial sum is zero, so we need a vector with all elements initialized to zero.
+    Vector128<int> sum = Vector128<int>.Zero;
+
+    // We need to obtain the reference to first value in the buffer, it's used later for loading vectors from memory.
+    ref int searchSpace = ref MemoryMarshal.GetReference(buffer);
+    // And an offset, that is going to be used by vectorized and scalar loops.
+    nuint elementOffset = 0;
+    // And the last valid offset from which we can load the values
+    nuint oneVectorAwayFromEnd = (nuint)(buffer.Length - Vector128<int>.Count);
+    for (; elementOffset <= oneVectorAwayFromEnd; elementOffset += (nuint)Vector128<int>.Count)
+    {
+        // We load a vector from given offset.
+        Vector128<int> loaded = Vector128.LoadUnsafe(ref searchSpace, elementOffset);
+        // We add 4 integers at a time:
+        sum += loaded;
+    }
+
+    // We sum all 4 integers from the vector to one
+    int result = Vector128.Sum(sum);
+
+    // And handle the remaining elements, in a non-vectorized way:
+    while (elementOffset < (nuint)buffer.Length)
+    {
+        result += buffer[(int)elementOffset];
+        elementOffset++;
+    }
+
+    return result;
+}
+```
+
+**Note:** Use `ref MemoryMarshal.GetReference(span)` instead of `ref span[0]` and `ref MemoryMarshal.GetArrayDataReference(array)` instead of `ref array[0]` to handle empty buffer scenarios (which would throw `IndexOutOfRangeException`). If the buffer is empty, these methods return a reference to the location where the 0th element would have been stored. Such a reference may or may not be null. You can use it for pinning but you must never dereference it.
+
+**Note:** The `GetReference` method has an overload that accepts a `ReadOnlySpan` and returns mutable reference. Please use it with caution!
+
+**Note:** Please keep in mind that `Vector128.Sum` is a static method. `Vectior128<T>` and `Vector256<T>` provide both instance and static methods (operators like `+` are just static methods in C#). `Vector128` and `Vector256` are non-generic static classes with static methods only. It's important to know about their existence when searching for methods.
+
+### Vectorized remainder handling
+
+Now imagine that we need to check whether the given buffer contains specific number. In this case, processing some values more than once is acceptable, we don't need to handle the remainder in a non-vectorized fashion.
```

Comment:
```suggestion
Now imagine that we need to check whether the given buffer contains specific number. In this case, processing some values more than once is acceptable, we don't need to handle the remainder in a non-vectorized fashion. This can be done for all idempotent operations (e.g. also copying).
```
To make this a bit more general?